src/Package/Impl/Sql/Publish/SqlPublishOptionsDialogViewModel.cs

```diff
@@ -0,0 +1,315 @@
+            IWritableSettingsStorage settingsStorage, IProjectConfigurationSettingsProvider pcsp) {
+            _coreShell = coreShell;
+            _pss = pss;
+            _settingsStorage = settingsStorage;
+            _pcsp = pcsp;
+
+            Settings = new SqlSProcPublishSettings(settingsStorage);
+
+            SelectCodePlacementMode();
+            SelectQuoteType();
+            SelectTargetType();
+        }
+
+        public void SaveSettings() {
+            Settings.Save(_settingsStorage);
+        }
+
+        public void UpdateState() {
+            TargetHasName = Settings.TargetType != PublishTargetType.Dacpac;
+            GenerateTable = Settings.CodePlacement == RCodePlacement.Table;
+            CanPublish = !GenerateTable || !string.IsNullOrEmpty(Settings.TableName);
+            if (Settings.TargetType != PublishTargetType.Dacpac) {
+                CanPublish &= Targets?.Count > 0;
+            }
+        }
+
+        private void PopulateTargets() {
+            switch (Settings.TargetType) {
+                case PublishTargetType.Dacpac:
+                    Targets = new List<string>();
+                    break;
+
+                case PublishTargetType.Project:
+                    PopulateProjectList();
+                    break;
+
+                case PublishTargetType.Database:
+                    InitializationTask = PopulateConnectionsListAsync();
+                    TargetHasName = false;
+                    break;
+            }
+        }
+
+        private Task PopulateConnectionsListAsync() {
+            return GetDatabaseConnectionsAsync((connections) => {
+                var index = -1;
+                _targetConnections = connections;
+                if (!string.IsNullOrEmpty(Settings.TargetDatabaseConnection)) {
+                    var indices = _targetConnections.IndexWhere(c => c.ConnectionString.EqualsIgnoreCase(Settings.TargetDatabaseConnection));
+                    index = indices.DefaultIfEmpty(-1).First();
+                }
+                Targets = _targetConnections.Select(c => c.Name).ToList();
+                ForceSelectedTargetIndexUpdate();
+                SelectedTargetIndex = index >= 0 ? index : 0;
+                UpdateState();
+            });
+        }
+
+        private void PopulateProjectList() {
+            _targetProjects = GetDatabaseProjects();
+            if (_targetProjects.Count > 0) {
+                var index = -1;
+                if (!string.IsNullOrEmpty(Settings.TargetProject)) {
+                    var indices = _targetProjects.IndexWhere(x => x.EqualsIgnoreCase(Settings.TargetProject));
+                    index = indices.DefaultIfEmpty(-1).First();
+                }
+                Targets = _targetProjects;
+                ForceSelectedTargetIndexUpdate();
+                SelectedTargetIndex = index >= 0 ? index : 0;
+            }
+        }
+
+        private void ForceSelectedTargetIndexUpdate() {
```

Comment:
When collection is replaced, selection needs to be rest. But I can do it in the Targets setter